enhancements/agent-installer/interactive-network-config.md

```diff
@@ -0,0 +1,238 @@
+* Make available all necessary information about the hardware and network
+  environemnt to do this configuration.
+* Provide an indication of when sufficient connectivity is or is not available.
+* Enable the user to debug connectivity problems by narrowing down the cause
+  (DNS, reachability, &c.).
+* Work for both the existing automation workflow and a future fully-interactive
+  workflow.
+
+### Non-Goals
+
+* Interactive (re)configuration of any other aspect of the installation.
+* Add interactive network configuration to CoreOS in general.
+* Allow interactive network configuration during IPI baremetal installs.
+
+## Proposal
+
+### Workflow Description
+
+The user will generate the agent ISO using `openshift-install` as usual. The
+`networkConfig` section in the `agent-config` (or the NMStateConfig ZTP
+manifests) may _or may not_ contain network configuration.
+
+The user will then boot the ISO. Any network configuration for each current
+host contained in the ISO will be applied at startup.
+
+Prior to the login prompt, a TUI will appear on the hardware console. It will
+display the current status of connectivity to the registry, which will be
+continuously updated. It will also give the user the option to interactively
+reconfigure the network, or accept the current configuration and exit.
+
+If there is connectivity available to the registry, the TUI will autmatically
+exit within 20s if there is no user interaction.
+
+Services that require pulling a container image will not start until the TUI
+exits.
+
+Once the TUI exits, the login prompt will be displayed on the console as usual.
+
+### API Extensions
+
+N/A
+
+### Implementation Details/Notes/Constraints
+
+The TUI will be implemented as a golang binary and delivered in a container
+added to the release payload. The actual network configuration will be done by
+through the existing `nmtui` binary, which is already present in CoreOS.
+
+To allow it to run before the network is configured, the TUI binary will be
+extracted from the release payload by the installer process, and embedded in
+the CoreOS ISO (which is also extracted from the release payload). Since even a
+modest golang binary is likely too large for the pre-allocated ignition area,
+this will be added to the ISO either as a separate volume mount or as an
+additional CPIO archive appended to a rewritten ignition volume. In PXE
+installs, it will be in an additional CPIO archive appended to the rootfs.
+
+The running network configuration is copied to disk using the `--copy-network`
+argument to `coreos-installer`, so no other changes are required to persist the
+desired configuration to the network.
+
+Access to the registry could be checked by simply telling podman to pull the
+release image. However, if this fails then ideally we would be able to identify
+the point where it stops working in the following steps:
+
+* DNS lookup
+* ping
+* HTTP
+* Registry login
+* Pulling the release image
```

Comment:
In a disconnected environment it could also be an issue with the ICSP configuration. Would it be a step to change to change registries.conf to fix mirror misconfiguration or will that require regenerating the ISO?

Reply:
I think that would require regeneration, since it's used in multiple places (not just `registries.conf`, but also in the ZTP manifests. Also, we want it to be the same across hosts.